functions/inputs/from_test.go

```diff
@@ -142,3 +147,226 @@ func TestFromOpSpec_BucketsAccessed(t *testing.T) {
+			// count      mean
+			//     \     /          count     mean
+			//      range       =>      \    /
+			//        |                  from
+			//       from
+			rules: []plan.Rule{&inputs.FromRangeTransformationRule{}},
+			before: &plantest.PhysicalPlanSpec{
+				Nodes: []plan.PlanNode{
+					plan.CreatePhysicalNode("from", from),
+					plan.CreatePhysicalNode("range", rangeWithBounds),
+					plan.CreatePhysicalNode("count", count),
+					plan.CreatePhysicalNode("yield0", yield("count")),
+					plan.CreatePhysicalNode("mean", mean),
+					plan.CreatePhysicalNode("yield1", yield("mean")),
+				},
+				Edges: [][2]int{
+					{0, 1},
+					{1, 2},
+					{2, 3},
+					{1, 4},
+					{4, 5},
+				},
+			},
+			after: &plantest.PhysicalPlanSpec{
+				Nodes: []plan.PlanNode{
+					plan.CreatePhysicalNode("merged_from_range", fromWithBounds),
+					plan.CreatePhysicalNode("count", count),
+					plan.CreatePhysicalNode("mean", mean),
+				},
+				Edges: [][2]int{
+					{0, 1},
+					{0, 2},
+				},
+			},
+		},
+		{
+			name: "cannot push range into from",
+			// range    count                                      range    count
+			//     \    /       =>   cannot push range into a   =>     \    /
+			//      from           from with multiple sucessors         from
+			rules: []plan.Rule{&inputs.FromRangeTransformationRule{}},
+			before: &plantest.PhysicalPlanSpec{
+				Nodes: []plan.PlanNode{
+					plan.CreatePhysicalNode("from", from),
+					plan.CreatePhysicalNode("range", rangeWithBounds),
+					plan.CreatePhysicalNode("yield0", yield("range")),
+					plan.CreatePhysicalNode("count", count),
+					plan.CreatePhysicalNode("yield1", yield("count")),
+				},
+				Edges: [][2]int{
+					{0, 1},
+					{1, 2},
+					{0, 3},
+					{3, 4},
+				},
+			},
+			after: &plantest.PhysicalPlanSpec{
+				Nodes: []plan.PlanNode{
+					plan.CreatePhysicalNode("from", from),
+					plan.CreatePhysicalNode("range", rangeWithBounds),
+					plan.CreatePhysicalNode("count", count),
+				},
+				Edges: [][2]int{
+					{0, 1},
+					{0, 2},
+				},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			before := plantest.CreatePhysicalPlanSpec(tc.before)
+			after := plantest.CreatePhysicalPlanSpec(tc.after)
+
+			physicalPlanner := plan.NewPhysicalPlanner(
+				plan.WithPhysicalRule(tc.rules...),
+			)
+
+			pp, err := physicalPlanner.Plan(before)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			want := make([]fromTestAttrs, 0)
+			after.BottomUpWalk(func(node plan.PlanNode) error {
+				want = append(want, fromTestAttrs{
+					ID:   node.ID(),
+					Spec: node.ProcedureSpec(),
+				})
+				return nil
+			})
+
+			got := make([]fromTestAttrs, 0)
+			pp.BottomUpWalk(func(node plan.PlanNode) error {
+				got = append(got, fromTestAttrs{
+					ID:   node.ID(),
+					Spec: node.ProcedureSpec(),
+				})
+				return nil
+			})
+
+			if !cmp.Equal(want, got, semantictest.CmpOptions...) {
```

Comment:
Is there any reason not to use helper methods in  `plan.plantest` here to compare the expected and  actual plans?

Reply:
So I'm not sure if we want to have one helper method for comparing an entire plan anymore. Especially for physical plans we will be adding statistics and costs, and we might want to just test for specific attributes of a plan instead of all attributes. What do you think?

Another reason I didn't use the test helper was now that `PlanNodes` have bounds, after planning the bounds will be calculated, but there's not really an easy way to construct a plan with bounds from a list of nodes and edges. In order to do that we would need to make all of the attributes of a `PlanNode` public.

So I guess what I'm trying to say is that I'm still not satisfied with testing. It's not as easy and straight forward as I want it to be.

Reply:
I am of the opinion that rules-testing should be as concise and easy to write and to read as possible.  I think specifying the input as Flux is nice.  Specifying the expected output using Go syntax has thus far been painful.  I think you're right that we probably want each test to lock down the specific attributes that the test is concerned with, and disregard others.  Maybe a DSL would help do this.  It's worth filing an issue an talking about it as a team.